### ISSUE TITLE
Braze plugin already receives in-app messages before we can handle them

### ISSUE THREAD

**Issue by JordyLangen** [braze-inc/braze-flutter-sdk#5]
This is our (simplified) setup in the app's main:

```dart
  print("creating braze plugin...");
  final braze = BrazePlugin()
  print("braze plugin created");
  ```
  
I've updated the `_handleBrazeData` in the `BrazePlugin` to print the incoming message. This is what we are seeing in the logs:

```
I/flutter ( 6072): creating braze plugin...
I/flutter ( 6072): In app message received: {"message":"Hello from the other side","click_action":"NONE","btns":[{"id":0,"text":"No","click_action":"NONE","bg_color":4294967295,"text_color":4279990479,"border_color":4279990479},{"id":1,"text":"Yes","click_action":"NONE","bg_color":4279990479,"text_color":4294967295,"border_color":4279990479}],"type":"MODAL","text_align_message":"CENTER","message_close":"SWIPE","header":"Hello","text_align_header":"CENTER","close_btn_color":4291085508,"bg_color":4294243575,"frame_color":3207803699,"text_color":4280624421,"header_text_color":4280624421,"trigger_id":"NjEzMWZlNWFjN2ViYjQ2MzE1YTgxOWY4XyRfbXY9NjEzMWZlNWFjN2ViYjQ2MzE1YTgxYTAyJnBpPWNtcA=="}
I/flutter ( 6072): Braze in-app message callback not present. Doing nothing.
I/flutter ( 6072): braze plugin created
```

How come the message is coming in before we can handle it? I think the fact dart being single-threaded is an issue here. This issue is happening on Android with a "normal" setup using a `braze.xml` config.

If this is an asynchronicity issue, maybe a solution that uses a queue will help, so the messages can be flushed when an in app message handler is set later on.

**Comment by hokstuff** [comment]
Hi @JordyLangen,

Thanks for filing this issue! It looks like the in-app message callback is being called immediately after initializing the `BrazePlugin`, but before actually setting the callback.

In your local project, would you be able to try updating the [`BrazePlugin` init code](https://github.com/braze-inc/braze-flutter-sdk/blob/487384e8eba9e81c1ac42a4475cd7ef6031f9c78/lib/braze_plugin.dart#L10) to allow for an optional `brazeInAppMessageHandler` parameter, and then pass in your callback? That way, the callback should be set synchronously during initialization and prevent it from being referenced it while it is `null`. If this all looks good, we will update the plugin to add this functionality.

Thank you!

**Comment by RtgrV** [comment]
@hokstuff 

As Jordy is currently out of the office, I would like to reply to the suggested change. If I understand correctly, you think that when we would change the initializer to:
```
  BrazePlugin({Function(BrazeInAppMessage)? brazeInAppMessageCallback}) {
    _channel.setMethodCallHandler(_handleBrazeData);
    
    if (brazeInAppMessageCallback != null) {
      setBrazeInAppMessageCallback(brazeInAppMessageCallback)
    }
  }
  
  void setBrazeInAppMessageCallback(Function(BrazeInAppMessage) callback) {
    _brazeInAppMessageHandler = callback;
  }
```

That would resolve it?

**Comment by JordyLangen** [comment]
Quickly shiming in 😄 

For sure providing the callback handler in the constructor would resolve the observed issue.
However, in our case, and I guess in most cases, we don't have the handler ready when creating the `Blaze` plugin instance.

I think the nicest solution, that hides the problem from the outside world, would be something like this:

```dart
  
  final missedInAppMessages = <BrazeInAppMessage>[];
  
  Future<dynamic> _handleBrazeData(MethodCall call) async {
    switch (call.method) {
      case "handleBrazeInAppMessage":
        final Map<dynamic, dynamic> argumentsMap = call.arguments;
        String? inAppMessageString = argumentsMap['inAppMessage'];
        final brazeInAppMessageHandler = this._brazeInAppMessageHandler;
        final inAppMessage = BrazeInAppMessage(inAppMessageString!);
        if (brazeInAppMessageHandler != null) {
          brazeInAppMessageHandler(inAppMessage);
        } else {
          missedInAppMessages.add(inAppMessage);
        }
        return new Future.value("");
      case "handleBrazeContentCards":
        ...
      default:
        print("Unknown method " + call.method + " called. Doing nothing.");
        return new Future.value("");
    }
  }

  void setBrazeInAppMessageCallback(Function(BrazeInAppMessage) callback) {
    _brazeInAppMessageHandler = callback;

    if (missedInAppMessages.length > 1) {
      // TODO: send all missed messages
    }
  }
```

This concept would then also work for any missed content cards messages.
Added benefit is that everyone who updates the plugin gets this fix for free, instead of adding the callback handler in the constructor.

**Comment by hokstuff** [comment]
Hi @JordyLangen and @RtgrV,

Thanks for providing the code snippets! After discussing with the team, we feel that the added complexity when trying to hide the problem from the outside world may lead to undesirable or unexpected behavior in some use cases. For instance, if there are multiple IAMs triggered before the callback is set, the callbacks will all get triggered at once although only the first is likely to display.

We feel it's best to keep the API straightforward to the integrator instead of trying to anticipate their expected use cases and solve the issues under the hood. This is why we are planning to allow optional parameters for these two callbacks in the constructor to allow opting-in to this feature. (Btw your snippet LGTM @RtgrV).

Let me know if this works with you, and we will work on creating this change. Thanks!

**Comment by JordyLangen** [comment]
@hokstuff thanks for your reply.

> For instance, if there are multiple IAMs triggered before the callback is set, the callbacks will all get triggered at once although only the first is likely to display.

Isn't this depending on how the consuming client implemented it?

>  instead of trying to anticipate their expected use cases

What do you mean by this? In this case it's an issue with the plugin, not an use case.

We are more then happy to provide PRs providing several fixes if you are willing to take them.

**Comment by hokstuff** [comment]
Hi @JordyLangen,

Thanks for submitting the PR! We discussed as a team and decided to go with a solution that encapsulates both your proposed solution as well as the extended constructor solution (with optional parameters). We'll also add documentation around this implementation detail to make sure integrators are aware of any changes.

Since we use a private repo, we'll bring in these proposed changes there and then give you a shout out once it is released.

Thank you again!

**Comment by hokstuff** [comment]
Hi @JordyLangen,

Thank you again for filing this issue and submitting a PR! We have released Flutter SDK [version 2.2.0](https://github.com/braze-inc/braze-flutter-sdk/releases/tag/2.2.0) which implements your suggested changes from your related PR.

Best wishes,
Daniel
